### PR TITLE
Lars

### PR DESCRIPTION
**Pull request by larsbjn**
* Generating rooms
* Game Constants
* Pathfinder
* CommandProcess
* Interact

### CODE REVIEW

File: src/hospitalgame/Game.java [hunk 1]
```diff
@@ -1,5 +1,15 @@
...
+import hospitalgame.NPC.Doctor;
+import hospitalgame.NPC.NPC;
+import hospitalgame.NPC.Porter;
+import hospitalgame.item.BloodBag;
+import hospitalgame.item.Item;
+import hospitalgame.item.ItemName;
+import hospitalgame.item.PowerUpItem;
+import java.util.ArrayList;
+import java.util.Random;
 import sun.reflect.generics.reflectiveObjects.NotImplementedException;
```
Comment (frederikrosenberg) [inline src/hospitalgame/Game.java]: unused import

File: src/hospitalgame/Game.java [hunk 2]
```diff
@@ -34,109 +45,163 @@
...
+     */
+    private Doctor doctor;
+    
+    /**
+     * Determines if the game is over
      */
-    //private ArrayList<NPC> NPCs;
+    private boolean gameOver = false;
+    
+    private static Game instance;
```
Comment (frederikrosenberg) [inline src/hospitalgame/Game.java]: Javadoc

File: src/hospitalgame/Game.java [hunk 3]
```diff
@@ -34,109 +45,163 @@
...
-    public Game() 
-    {
-        createRooms();
+    private Game() {
+        map = new Map();
+        createRooms(1);
         parser = new Parser();
     }
     
+    public static Game getGameInstance() {
```
Comment (frederikrosenberg) [inline src/hospitalgame/Game.java]: Javadoc

File: src/hospitalgame/Game.java [hunk 4]
```diff
@@ -34,109 +45,163 @@
...
+        if(instance == null) {
+            instance = new Game();
+        }
+        return instance;
+    }
+    
+    public void setGameOver() {
```
Comment (frederikrosenberg) [inline src/hospitalgame/Game.java]: Javadoc

File: src/hospitalgame/Game.java [hunk 5]
```diff
@@ -34,109 +45,163 @@
...
-        currentRoom = outside;
+    private void createRooms(int numberOfRooms) {
+        // Gets all the bloodtypes into a BloodType array.
+        BloodType[] bloodType = BloodType.values();
+        // Creates a new Random object.
+        Random random = new Random();
+        // Random picks the players bloodtype.
+        BloodType playerBloodType = bloodType[random.nextInt(bloodType.length)];
+        // Initialize a new player object.
+        player = new Player(playerBloodType, GameConstants.PLAYER_BLOODRATE, GameConstants.PLAYER_BLOOD_AMOUNT, "Lars");
```
Comment (frederikrosenberg) [inline src/hospitalgame/Game.java]: Player name should not be lars but need to be something else

File: src/hospitalgame/Game.java [hunk 6]
```diff
@@ -34,109 +45,163 @@
...
 
         if (commandWord == CommandWord.HELP) {
             printHelp();
-        }
-        else if (commandWord == CommandWord.GO) {
-            goRoom(command);
-        }
-        else if (commandWord == CommandWord.QUIT) {
+        } else if (commandWord == CommandWord.GO) {
+            if(player.goRoom(command)) {
```
Comment (jazzmano) [inline src/hospitalgame/Game.java]: Change to switch statement

File: src/hospitalgame/GameConstants.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+ * and open the template in the editor.
+ */
+package hospitalgame;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+
+/**
+ *
+ * @author larsjorgensen
```
Comment (AndreasBoegh) [inline src/hospitalgame/GameConstants.java]: kun 1 author???

File: src/hospitalgame/Map.java [hunk 8]
```diff
@@ -15,9 +24,126 @@
...
+        // Sets the doctors room in the Porter object.
+        porter.setEndRoom(doctor.getCurrentRoom());
+        
+        // Sets the start room to the first free room.
+        Room startRoom = freeRooms.get(0);
+        // Creates the queue where all the rooms that needs to be processed is stored.
+        Queue<Room> roomsToProcess = new LinkedList<>();
+        // Adds the first free room to the queue.
+        roomsToProcess.add(freeRooms.remove(0));
+        // Aslong as the queue is not empty.
```
Comment (jazzmano) [inline src/hospitalgame/Map.java]: As long*

File: src/hospitalgame/Map.java [hunk 9]
```diff
@@ -15,9 +24,126 @@
...
+        while (!roomsToProcess.isEmpty()) {
+            // Gets the next room in the queue.
+            Room currentRoom = roomsToProcess.poll();
+            int i = 0;
+            // Generates a random number of exits the room have.
+            int exitCount = (int)(Math.random() * 4) + 1;
+            // Aslong as there is free rooms left and the room needs more exits.
```
Comment (jazzmano) [inline src/hospitalgame/Map.java]: As long*

File: src/hospitalgame/Map.java [hunk 10]
```diff
@@ -15,9 +24,126 @@
...
+            }  
+        }
+        // returns the start room.
+        return startRoom;
+    }
+    
+    /**
+     * Creates the rooms.
+     * @param roomCount The count of how many rooms that will be generated.
+     * @return A arraylist of all the generated rooms.
```
Comment (jazzmano) [inline src/hospitalgame/Map.java]: An*

File: src/hospitalgame/Room.java [hunk 11]
```diff
@@ -42,6 +42,7 @@ public Room(String description)
     {
         this.description = description;
         exits = new HashMap<String, Room>();
+        itemArray = new ArrayList<>();
     }
```
Comment (frederikrosenberg) [inline src/hospitalgame/Room.java:45]: Format please

**Review by frederikrosenberg — CHANGES_REQUESTED** [review]
small changes + javadoc

File: src/hospitalgame/Map.java [hunk 12]
```diff
@@ -15,9 +24,126 @@
     /**
      * Generate the rooms for the game.
```
Comment (AndreasBoegh) [inline src/hospitalgame/Map.java]: stavefejl

File: src/hospitalgame/GameConstants.java [hunk 13]
```diff
@@ -0,0 +1,27 @@
+/*
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
```
Comment (AndreasBoegh) [inline src/hospitalgame/GameConstants.java]: Fjern license header

File: src/hospitalgame/Game.java [hunk 14]
```diff
@@ -145,36 +210,11 @@ private void printHelp()
...
-            System.out.println("There is no door!");
-        }
-        else {
-            currentRoom = nextRoom;
-            System.out.println(currentRoom.getLongDescription());
-        }
-    }
-
-    /**
      * Checking if the player want to quit.
```
Comment (AndreasBoegh) [inline src/hospitalgame/Game.java]: stavefejl

File: src/hospitalgame/Game.java [hunk 15]
```diff
@@ -34,109 +45,163 @@
...
+    public void play() {            
         printWelcome();
 
         boolean finished = false;
         while (! finished) {
             Command command = parser.getCommand();
-            finished = processCommand(command);
+            finished = processCommand(command) || gameOver;
         }
         System.out.println("Thank you for playing.  Good bye.");
```
Comment (AndreasBoegh) [inline src/hospitalgame/Game.java]: Stavefejl

Comment (AndreasBoegh) [reply src/hospitalgame/Game.java]: goodbye i 1 ord

**Review by AndreasBoegh — CHANGES_REQUESTED** [review]

**Review by jazzmano — APPROVED** [review]

**Review by frederikrosenberg — APPROVED** [review]

**Review by AndreasBoegh — APPROVED** [review]

**Pull request merged**